edgelet/edgelet-kube/src/runtime.rs

```diff
@@ -4,30 +4,36 @@ use std::cell::RefCell;
 use std::sync::{Arc, Mutex};
 use std::time::Duration;
 
-use crate::constants::*;
-use crate::convert::{auth_to_image_pull_secret, pod_to_module, spec_to_deployment};
-use crate::error::{Error, ErrorKind, Result};
-use crate::module::KubeModule;
-use edgelet_core::{
-    LogOptions, ModuleRegistry, ModuleRuntime, ModuleRuntimeState, ModuleSpec, ModuleTop,
-    RuntimeOperation, SystemInfo,
-};
-use edgelet_docker::DockerConfig;
-use edgelet_utils::{ensure_not_empty_with_context, sanitize_dns_label};
 use failure::Fail;
 use futures::future::Either;
 use futures::prelude::*;
 use futures::{future, stream, Async, Future, Stream};
 use hyper::client::HttpConnector;
 use hyper::service::Service;
-use hyper::{Body, Chunk as HyperChunk};
+use hyper::{header, Body, Chunk as HyperChunk, Request};
 use hyper_tls::HttpsConnector;
-use kube_client::{get_config, Client as KubeClient, HttpClient, ValueToken};
+use log::Level;
 use url::Url;
 
+use edgelet_core::{
+    AuthId, Authenticator, LogOptions, ModuleRegistry, ModuleRuntime, ModuleRuntimeState,
+    ModuleSpec, RuntimeOperation, SystemInfo,
+};
+use edgelet_docker::DockerConfig;
+use edgelet_utils::{ensure_not_empty_with_context, log_failure, sanitize_dns_label};
+use kube_client::HttpClient;
+use kube_client::{
+    get_config, Client as KubeClient, Error as KubeClientError, TokenSource, ValueToken,
+};
+
+use crate::constants::*;
+use crate::convert::{auth_to_image_pull_secret, pod_to_module, spec_to_deployment};
+use crate::error::{Error, ErrorKind, Result};
+use crate::module::KubeModule;
+
 #[derive(Clone)]
-pub struct KubeModuleRuntime<S> {
-    client: Arc<Mutex<RefCell<KubeClient<ValueToken, S>>>>,
+pub struct KubeModuleRuntime<T: Clone, S> {
```

Comment:
It should be possible to not have to require `Clone` on `T` here. You'll need to remove the bounds on the `Client` struct and the `Config` struct in `kube-client` for this though.

Reply:
Nice!